snorkeling_helper/label_functions/knowledge_bases/hetionet_knowledge_base_generator.py

```diff
@@ -0,0 +1,380 @@
+username = "danich1"
+password = "snorkel"
+dbname = "pubmed_central_db"
+
+# Path subject to change for different os
+database_str = (
+    f"postgresql+psycopg2://{username}:{password}@/{dbname}?host=/var/run/postgresql"
+)
+conn = create_engine(database_str)
+# -
+
+# # Disease associates Gene
+
+# ## Disease and Gene Info URLs for Hetionet
+
+# +
+disease_url = "https://raw.githubusercontent.com/dhimmel/disease-ontology/052ffcc960f5897a0575f5feff904ca84b7d2c1d/data/xrefs-prop-slim.tsv"
+gene_url = "https://raw.githubusercontent.com/dhimmel/entrez-gene/a7362748a34211e5df6f2d185bb3246279760546/data/genes-human.tsv"
+dag_url = "https://github.com/dhimmel/integrate/raw/93feba1765fbcd76fd79e22f25121f5399629148/compile/DaG-association.tsv"
+
+# Not used but url for disease up/down regulates gene
+# Leaving for future work if someone picks this project up
+drg_url = "https://raw.githubusercontent.com/dhimmel/stargeo/08b126cc1f93660d17893c4a3358d3776e35fd84/data/diffex.tsv"
+# -
+
+# ## Load DataFrames for each URL
+
+disease_ontology_df = (
+    pd.read_csv(disease_url, sep="\t")
+    >> ply.distinct(["doid_code", "doid_name"])
+    >> ply.rename(doid_id="doid_code")
+    >> ply.define(merge_key=1)
+)
+disease_ontology_df >> ply.slice_rows(5)
+
+entrez_gene_df = (
+    pd.read_csv(gene_url, sep="\t")
+    >> ply.rename(dict(entrez_gene_id="GeneID", gene_symbol="Symbol"))
+    >> ply.define(merge_key=1)
+)
+entrez_gene_df >> ply.slice_rows(5)
+
+disease_gene_map_df = (
+    entrez_gene_df
+    >> ply.select("entrez_gene_id", "gene_symbol", "merge_key")
+    >> ply.inner_join(
+        disease_ontology_df >> ply.select("doid_id", "doid_name", "merge_key"),
+        on="merge_key",
+    )
+    >> ply.select("-merge_key")
+)
+disease_gene_map_df >> ply.slice_rows(5)
+
+hetionet_dag_df = pd.read_csv(
+    dag_url, sep="\t", dtype={"entrez_gene_id": int}
+) >> ply.define(merge_key=1)
+hetionet_dag_df >> ply.slice_rows(5)
+
+query = """
+SELECT "disease_cid" AS doid_id, "gene_cid" AS  entrez_gene_id, count(*) AS n_sentences
+FROM disease_gene
+GROUP BY "disease_cid", "gene_cid"
+"""
+disease_gene_sentence_df = (
+    pd.read_sql(query, conn)
+    >> ply_tdy.separate_rows("entrez_gene_id", sep=";")
+    >> ply.call(".astype", {"entrez_gene_id": int})
+)
+disease_gene_sentence_df >> ply.slice_rows(5)
+
+# ## Merge all dataframes into One
+
+disease_gene_associations_df = (
+    disease_gene_map_df
+    >> ply.left_join(
+        hetionet_dag_df >> ply.select("doid_id", "entrez_gene_id", "sources"),
+        on=["doid_id", "entrez_gene_id"],
+    )
+    >> ply.left_join(disease_gene_sentence_df, on=["doid_id", "entrez_gene_id"])
+    >> ply.call(".fillna", {"n_sentences": 0})
+    >> ply.call(".astype", {"n_sentences": int})
+    >> ply.define(
+        hetionet="sources.notnull().astype(int)",
+        has_sentence="(n_sentences > 0).astype(int)",
+    )
+)
+(disease_gene_associations_df >> ply.slice_rows(5) >> ply.call(".transpose"))
```

Comment:
Looks like the dataframe does not have column names just indexes? not sure if it is important though.

Reply:
In this example shown where `n_sentences` is 4 for _A1BG_ and hematologic cancer, I guess it means you have four sentences basically mentioning this relationship, right? If so, how do you know the direction of the association? Maybe it says "A1BG show no association with hematologic cancer". I guess the model detects that but just curious.

Reply:
>Looks like the dataframe does not have column names just indexes? not sure if it is important though.

Ah the dataframe does have columns. I'm just transposing it, so it is easier to read. Columns are the indicies.

>In this example shown where n_sentences is 4 for A1BG and hematologic cancer, I guess it means you have four sentences basically mentioning this relationship, right? If so, how do you know the direction of the association? Maybe it says "A1BG show no association with hematologic cancer". I guess the model detects that but just curious.

Fantastic question. Determining relationship direction can be tricky as the disease could be associated with the gene or vise versa. Unfortunately, that challenge is mainly for the knowledge graph itself rather than for this NLP project. If i recall correctly Heionet defines Disease associates Gene as the disease -> gene. An extension for my current project would be to detect the direction of relationships based on the sentences. I wouldn't be surprised if there are papers out there that did this sort of task before; however, to reiterate relationship direction is inherited from [Hetionet's schema](https://het.io/about/metagraph.png).